Specifying culture to use when serializing value types in MonoGame

Lucasharskamp opened the pull request:
This PR contains a few solutions to (potential) issues with serializing value types, as discussed in detail in https://github.com/MonoGame/MonoGame/issues/8923

1) ValueTypes (like `Quaternion`) now have an overload to their ToString() methods where one can provide their desired `IFormatProvider`
2) All existing usages of ToString, String.Format or other features that may have different behavior based on their culture are now specified as using `CultureInfo.InvariantCulture`, to ensure we get the same behavior regardless of whatever unknown culture rules are in use on someone's machine. 
3) StringBuilder methods where a single char was appended are now done as adding a char rather than a string of 1 char in length.

Lucasharskamp committed:
Improved serialization for value types to ensure programmers can apply their preferred culture formatting.
Fixed potential issues on exotic computers

Lucasharskamp committed:
Merge branch 'MonoGame:develop' into minor/dataserialization

ThomasFOG commented:
I'll continue reviewing this later, but it's a bit hard to read because there are changes of different kind which aren't useful/not desirable. It feels like you didn't try to target specific parts and just batch replaced things without checking if it's useful or breaking.

Before trying to fix any of that, we need to make sure that #8923 _is_ an issue with reproducible use cases.

Lucasharskamp commented:
> I'll continue reviewing this later, but it's a bit hard to read because there are changes of different kind which aren't useful/not desirable. It feels like you didn't try to target specific parts and just batch replaced things without checking if it's useful or breaking.
> 
> Before trying to fix any of that, we need to make sure that #8923 _is_ an issue with reproducible use cases.

I am having problems with that, hence why I made the PR. In my case, it's about logging diagnostic data for debugging purposes.
Here's an example:

```
CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;
var foo = new Vector2(100.4f, 50.5f);
Console.WriteLine(foo.ToString()); // { X:100.4, Y:50.5 }
Console.Writeline(foo.DebugDisplayString); // 100.4 50.5
CultureInfo.CurrentCulture = new CultureInfo("de-DE"); // German. Also applies to Dutch, French, etc.
Console.WriteLine(foo.ToString()); // { X:100,4, Y:50,5 }
Console.WriteLine(foo.DebugDisplayString); // 100,5  50,2
```

Of course there are workarounds (serializing fields by hand or using stuff like Newtonsoft.Json or System.Text.Json that does this for you), but properties like `DebugDisplayString` should simply be Culture agnostic out-of-the-box, and ToString should allow the programmer to specify their desired Culture to utilise (as .NET does for most default value types)

In the other examples it's about ensuring bugs related to the machine the code runs on cannot pop up; I've learned that one shouldn't make assumptions about the culture of the machine .NET code runs on, and presume it's something exotic, which can cause different behavior to pop up nobody anticipated.  So these are more ''just to be safe''.

SimonDarksideJ commented:
Any progress @Lucasharskamp on defining a FAILING Unit test for https://github.com/MonoGame/MonoGame/issues/8923

Before we start on a PR to fix the issues raised by the Unit test.  We cannot fix something unless we have a testable identification of the problem.

Lucasharskamp commented:
``

> Any progress @Lucasharskamp on defining a FAILING Unit test for #8923
> 
> Before we start on a PR to fix the issues raised by the Unit test. We cannot fix something unless we have a testable identification of the problem.

I'm creating a UnitTest that does the following for every Value type:
- Set different distinct CultureInfo's, serialize that way, see if it serializes as expected (string comparison - using en-US, en-GB, fr-FR, de-DE, nl-NL, it-IT,  ja-JP, zh-Hans)
- Set CultureInfo explicitly when serializing
- Ensure DebugDisplayString is always equal regardless of machine/set culture.
- For ''failure'': serialize a value type 2 times (in different set cultures) and it fails because the output is not equal. 

**Edit**: It's a lot of items to go through, I'll have it finished before the 22nd.

SimonDarksideJ commented:
Hay @Lucasharskamp , just checking in on how the tests were going.  Would be great to get this included in the 3.8.5 timelines, but no immediate rush

ThomasFOG commented:
Could you remove culture changes in places where it has nothing to do with printing data?

Lucasharskamp committed:
Added unittests for the serialization
Added vector2/vector3/vector4 serialization options, and for the packed types that utilise them.

Lucasharskamp committed:
Merge branch 'minor/dataserialization' of https://github.com/Lucasharskamp/MonoGame into minor/dataserialization

Lucasharskamp commented:
> Hay @Lucasharskamp , just checking in on how the tests were going. Would be great to get this included in the 3.8.5 timelines, but no immediate rush

Done. I've only added a few test - I could add one for each type, but that would make it a "repeated method" about two dozen times over, only changing the type to alter, without adding anything new. If preferred, I could of course add those anyway.


> Could you remove culture changes in places where it has nothing to do with printing data?

Convert.ChangeType can lead to unexpected issues depending on the Culture present, as it under the hood utilises _IConvertible_.  A few generic examples of this happening are included in the new unittests (method SerializationTest.cs -> TestChangeType). I doubt existing MonoGame implementations could lead to such issues occurring, but I wouldn't be surprised either if they did; it's a "better be safe than sorry" for edge cases with rare cultures.  
See https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Convert.cs,3bcca7a9bda4114e

The other places where changes were made have similar issues.

MonogameGuy commented:
The purposed change could break times, dates, fonts, and monetary units. 

If you want a `ToString()`  formatted a certain way, pass the Culture as intended in your game. The base `ToString()` was not meant for end user consumption. It is primarily a debugging aid, but when used in a game, it is not expected to vary based on culture, without the programmer asking it to.

The `Date` format alone shows the danger of mixing formats in strings or constants. What is the day after 11/11/xx: 12/11/xx or 11/12/xx? 

For my games, It's about making the runtime behavior deterministic across all locales. 
Otherwise it is a debugging nightmare. 

Please add a Culture `override` method where not provided in the framework:
```
public override string ToString()
        {
            return X.ToString();
        }
 public override string ToString(CultureInfo cultureInfo)
        {
            return X.ToString("X", cultureInfo);
        }
```
Adding a culture‑aware `ToString(CultureInfo cultureInfo)` is absolutely the right thing to do when you care about determinism and locale independence. The pattern above gives you both safety (invariant output) and flexibility (culture‑specific overloads).

What may seem like an innocuous simple change can be quite harmful to programs dependent upon the existing behavior.
